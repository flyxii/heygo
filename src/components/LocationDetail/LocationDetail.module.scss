@import "share/styles/base";

.wrapper {
  width: 100%;
  height: auto;
  box-sizing: border-box;
  padding: 0em;
  margin-top: 3em;

  @media only screen and (min-width: $md) {
    margin-top: 4em;
  }
}

.title{
  font-size: 2em;
  font-weight: 600;
  color: $darkBlue;
  margin-bottom: 1em;
}
