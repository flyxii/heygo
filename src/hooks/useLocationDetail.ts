/**
 * The query to fetch the location detail
 */

import { useQuery, QueryObserverResult } from "react-query";

type LocationDetail = {
  name: string;
  id: number;
  latitude: number;
  longitude: number;
  message?: string;
};

const getLocationDetail = async (id: string) => {
  const response = await fetch(
    `https://code-challenge-backend.herokuapp.com/locations/${id}`
  );
  const data = await response.json();
  return data;
};

export const useLocationDetail = (
  id: string
): QueryObserverResult<LocationDetail, unknown> => {
  return useQuery(["locationDetail", id], () => getLocationDetail(id));
};
