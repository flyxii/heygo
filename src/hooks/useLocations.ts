/**
 * The query to fetch the location list
 */

import { useQuery, QueryObserverResult } from "react-query";

type Locations = {
  name: string;
  id: number;
};

const getLocations = async (query: string) => {
  const response = await fetch(
    `https://code-challenge-backend.herokuapp.com/locations?q=${query}`
  );
  const data = await response.json();
  return data;
};

export const useLocations = (
  query: string
): QueryObserverResult<Locations[], unknown> => {
  return useQuery(["locations", query], () => getLocations(query));
};
