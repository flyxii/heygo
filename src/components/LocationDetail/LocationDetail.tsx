/**
 * Component to display the detail of single location
 * @querystring {id} the id of a single location
 */

import React from "react";
import { useParams, Redirect } from "react-router-dom";
import { useLocationDetail } from "hooks/useLocationDetail";
import Map from "./Map";
import styles from "./LocationDetail.module.scss";

type QueryString = {
  id: string;
};

const LocationDetail: React.FC = () => {
  /**
   * Get the id of a single location
   */
  const { id } = useParams() as QueryString;

  /**
   * Get the detail of a single location
   */
  const { isLoading, error, data } = useLocationDetail(id);

  /**
   * Display error message if fetch return error
   */
  if (error)
    return (
      <div className={styles.wrapper} data-testid="error">
        Error fetching data...
      </div>
    );

  /**
   * Display loading message before fetch complete
   */
  if (isLoading)
    return (
      <div className={styles.wrapper} data-testid="loading">
        We are working hard to get the data...
      </div>
    );

  /**
   * Redirect to home page if no data return from api call or location not found
   * This approach is just a demonstration for the coding test only, but not simulate a real world scenario
   */
  if (!data || data.message === "Not found") return <Redirect to="/" />;

  return (
    <div className={styles.wrapper}>
      <div className={styles.title}>{data.name}</div>
      <Map lat={data.latitude} lng={data.longitude} />
    </div>
  );
};

export default LocationDetail;
