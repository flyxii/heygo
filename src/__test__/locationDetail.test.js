import * as React from "react";
import { render } from "@testing-library/react";
import { Route } from "react-router-dom";
import LocationDetail from "components/LocationDetail";
import { useLocationDetail } from "hooks/useLocationDetail";

const mockUseLocationDetail = useLocationDetail; 
jest.mock("hooks/useLocationDetail");

/*jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest
    .fn()
    .mockReturnValue({ environment: "dev", service: "fakeService" }),
}));*/

describe("<LocationDetail />", () => {
  beforeEach(() => {
    mockUseLocationDetail.mockImplementation(() => ({
      isLoading: true,
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Renders without crashing", () => {
    const { getByTestId, asFragment } = renderWithRouter(
      () => (
        <Route path="/location/:id">
          <LocationDetail />
        </Route>
      ),
      "/location/1234"
    );

    const loading = getByTestId("loading");

    expect(loading).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("Display error message correctly", () => {
    mockUseLocationDetail.mockImplementation(() => ({
      isLoading: false,
      error: true,
    }));

    const { getByTestId, asFragment } = renderWithRouter(
      () => (
        <Route path="/location/:id">
          <LocationDetail />
        </Route>
      ),
      "/location/1234"
    );

    const error = getByTestId("error");

    expect(error).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("Redirect to / if no data return", () => {
    const mockData = { message: "Not found" };
    mockUseLocationDetail.mockImplementation(() => ({
      isLoading: false,
      data: mockData,
    }));

    const { container } = renderWithRouter(
      () => (
        <Route path="/location/:id">
          <LocationDetail />
        </Route>
      ),
      "/location/1234"
    );

    expect(container.innerHTML).toEqual(expect.stringContaining(""));
  });

  it("Displays data correctly", () => {
    const mockData = {
        name: "mockName",
        id: 1234,
        latitude: 1234,
        longitude: 1234,
      };
    mockUseLocationDetail.mockImplementation(() => ({
      isLoading: false,
      data: mockData,
    }));

    const { getByText, getByTestId, asFragment } = renderWithRouter(
      () => (
        <Route path="/location/:id">
          <LocationDetail />
        </Route>
      ),
      "/location/1234"
    );

    const location = getByText("mockName");
    const map = getByTestId("map");

    expect(location).toBeInTheDocument();
    expect(map).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
