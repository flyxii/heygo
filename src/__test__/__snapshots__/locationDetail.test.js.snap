// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationDetail /> Display error message correctly 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
    data-testid="error"
  >
    Error fetching data...
  </div>
</DocumentFragment>
`;

exports[`<LocationDetail /> Displays data correctly 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="title"
    >
      mockName
    </div>
    <div
      class="wrapper"
      data-testid="map"
    >
      <div
        style="width: 100%; height: 100%; margin: 0px; padding: 0px; position: relative;"
      >
        <div
          style="width: 100%; height: 100%; left: 0px; top: 0px; margin: 0px; padding: 0px; position: absolute;"
        />
        <div
          style="width: 50%; height: 50%; left: 50%; top: 50%; margin: 0px; padding: 0px; position: absolute;"
        >
          <div
            style="width: 100%; height: 100%; left: 0px; top: 0px; margin: 0px; padding: 0px; position: absolute;"
          >
            <div
              style="width: 0px; height: 0px; background-color: transparent; position: absolute;"
            >
              <div>
                <img
                  alt="Pins"
                  src="Pins.svg"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<LocationDetail /> Renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
    data-testid="loading"
  >
    We are working hard to get the data...
  </div>
</DocumentFragment>
`;
